cloudant-client/src/test/java/com/cloudant/tests/ReplicationMockTest.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.cloudant.tests;
+
+import static com.cloudant.tests.HttpTest.takeN;
+import static com.cloudant.tests.util.MockWebServerResources.EXPECTED_OK_COOKIE;
+import static com.cloudant.tests.util.MockWebServerResources.EXPECTED_OK_COOKIE_2;
+import static com.cloudant.tests.util.MockWebServerResources.JSON_OK;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+
+import com.cloudant.client.api.CloudantClient;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+
+import java.nio.charset.Charset;
+
+/**
+ * Created by samsmith on 09/03/2018.
+ */
+
+public class ReplicationMockTest {
+
+    @Rule
+    public MockWebServer mockWebServer = new MockWebServer();
+
+    final private static String authJson = "\"auth\":{\"iam\":{\"api_key\":\"";
+    final private static String sourceDbUrl = "https://foo.cloudant.com/source";
+    final private static String targetDbUrl = "https://bar.cloudant.com/target";
+
+    @Test
+    public void createAndTriggerReplicationWithNoIamAuth() throws Exception {
+        mockWebServer.enqueue(JSON_OK);
+
+        CloudantClient c = CloudantClientHelper.newMockWebServerClientBuilder(mockWebServer)
+                .build();
+
+        c.replication()
+                .source(sourceDbUrl)
+                .target(targetDbUrl)
+                .trigger();
+
+        RecordedRequest[] requests = takeN(mockWebServer, 1);
+
+        assertEquals(requests[0].getPath(), "/_replicate");
+
+        String body = requests[0].getBody().readString(Charset.forName("UTF-8"));
+
+        assertThat("The replication document should contain the correct source",
+                body, containsString("\"source\":\"" + sourceDbUrl + "\""));
+
+        assertThat("The replication document should contain the correct target",
+                body, containsString("\"target\":\"" + targetDbUrl + "\""));
+    }
+
+    @Test
+    public void createAndTriggerReplicationWithSourceIamAuth() throws Exception {
+        mockWebServer.enqueue(JSON_OK);
+
+        CloudantClient c = CloudantClientHelper.newMockWebServerClientBuilder(mockWebServer)
+                .build();
+
+        c.replication()
+                .source(sourceDbUrl)
+                .sourceIamApiKey(EXPECTED_OK_COOKIE)
```

Comment:
I think we should use something that has the format of the IAM API key to help us identify any possible issues with encoding etc
Looking at it the `com.cloudant.tests.HttpIamTest#iamApiKey` would probably be an appropriate thing to make public and share between tests. Of course the value would need updating to something useful instead of the current `"iam"` string value.

Reply:
The `EXPECTED_OK_COOKIE` is imported from ` MockWebServerResources`.

It's a bit confusing as it is an IAM API token. Badly named perhaps? There is an `IAM_TOKEN` too but that's the whole JSON response.

See [here](https://github.com/cloudant/java-cloudant/blob/master/cloudant-client/src/test/java/com/cloudant/tests/util/MockWebServerResources.java#L46).

Reply:
The `EXPECTED_OK_COOKIE` is representative of the cookie response to `_session` or `_iam_session` requests. The API key is the thing that gets sent to the IAM server to get an IAM token to send to `_iam_session`. The thing that gets put in the document should be an IAM API key, which is neither the cookie nor the token.

That said, it makes little difference here, they are both the cookie and an IAM API key use Base64 (URL style) strings (i.e `[A-Z][a-z][0-9][-_]`). AFAICT an IAM API key is 45 characters while the cookie is longer, but I think on further investigation I'm not worried about other characters being encoded. It might be beneficial to have something like
```java
// IAM API keys are also Base64 URL strings so use the cookie as an example key
public static final String IAM_API_KEY = EXPECTED_OK_COOKIE;
```
and then using that constant in the tests to avoid it looking like the replicator document should contain a cookie.